### PR TITLE
Add unit test on MacOS using Github Actions

### PR DESCRIPTION
**Pull request by alalazo**
refers #13485

### CODE REVIEW

**Comment by alalazo** [comment]
Hmmm, this time MacOS tests seem stuck. Trying to close and reopen. Note this is a rebase of #13826 pushed to `spack/spack`

**Comment by alalazo** [comment]
MacOS report uploaded!

**Comment by gartung** [comment]
@alalazo see https://github.com/spack/spack/pull/14236
Spack test will need to be run with
```
spack test --basetemp /tmp/spack
```
on macOS once gnupg2 is found correctly.

**Comment by alalazo** [comment]
There's another thing we need to understand before merging this PR:
> With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is **triggered from a forked repository.**

The sentence above can be found [here](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets) and it's not completely clear to me what does the emphasized part mean. 

The previous PR #13826 couldn't upload to `codecov` because it couldn't find the appropriate secret token. That's reasonable, since from a forked repository I submitted a PR that was trying to use a secret of this repository. I wonder though what happens if we were to merge this in the case of a normal PR that doesn't touch the workflow file but comes from a forked repository - which is fairly normal on Github. If they can't use the secret token to upload to codecov it would be a major pain point.

**Comment by alalazo** [comment]
To follow on the previous comments we might have an answer:
- https://github.com/codecov/codecov-action/issues/43#issuecomment-567889465
- https://github.com/codecov/codecov-action/issues/44#issuecomment-565447142

It seems this PR needs to wait for early 2020 to be merged, see https://github.com/codecov/codecov-action/issues/44#issuecomment-566695559. Before a tokenless upload to `codecov` is implemented all the forks need to store the same secret as the upstream repo in order to upload coverage which is definitely less than optimal.

**Comment by alalazo** [comment]
Note there's an [issue open for passing secrets to forks](https://github.community/t5/GitHub-Actions/Make-secrets-available-to-builds-of-forks/td-p/30678). It's marked as "solved" but really means "won't fix".

**Comment by alalazo** [comment]
@tgamblin For convenience a few references as we were discussing this topic offline:
- [Permissions given by the `GITHUB_TOKEN`](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token#permissions-for-the-github_token)
- [Permissions required to Apps to make API calls](https://developer.github.com/v3/apps/permissions/)

`GITHUB_TOKEN` is the only secret passed to forked repos and it basically gives `read/write` permissions on a lot of stuff if it comes from THIS repo and only `read` if it comes from a fork. Note that **we are not using it in this workflow** and my understanding is that it will be useful for us only if we start having workflows that e.g. merge or label issues and PRs automatically (see for instance what you can do with the "contents" permissions).

Currently the only way to have this PR work seamlessly also for forks is to expose the token from `codecov` in the YAML file - since there's no option at the moment to authenticate with `codecov` from a fork and preserve our secret `codecov` token at the same time. The risk is that anybody will in principle be able to upload `xml` reports to our `codecov`. Let me know if you want to proceed with that or if we should wait until the `codecov` guys work on a tokenless integration with Github Actions (the same way they are doing with Travis).

**Comment by alalazo** [comment]
I tried to update this PR since the homepage of [codecov/codecov-action](https://github.com/codecov/codecov-action) now says:

> **For public repositories, no token is needed**

We have another issue to investigate though. It seems that due to a setup specific to Github Actions tests that use `capfd` to capture output fail and apparently all the output is sent directly to `stdout` and `stderr` instead of being captured.

File: .github/workflows/macos_unit_tests.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  build:
+
+    runs-on: macos-latest
+    strategy:
+      max-parallel: 4
+      matrix:
+        python-version: [3.7]
+
+    steps:
+    - uses: actions/checkout@v1
```
Comment (ax3l) [inline .github/workflows/macos_unit_tests.yaml]: Use already the newer version, v1 had some issues recently.
```suggestion
    - uses: actions/checkout@v2
```

File: .github/workflows/macos_unit_tests.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+          pip install --upgrade setuptools
+          pip install --upgrade codecov coverage==4.5.4
+          pip install --upgrade flake8
+          pip install --upgrade pep8-naming
+    - name: Setup Homebrew packages
+      run: |
+        brew update
+        brew upgrade
+        brew install gcc gnupg2 dash kcov
+    - name: Run unit tests
```
Comment (ax3l) [inline .github/workflows/macos_unit_tests.yaml:41]: Maybe try: https://github.com/kelaberetiv/TagUI/issues/86
```suggestion
      brew uninstall openssl
    - name: Run unit tests
```

File: .github/workflows/macos_unit_tests.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          pip install --upgrade coverage
+          pip install --upgrade setuptools
+          pip install --upgrade codecov coverage==4.5.4
+          pip install --upgrade flake8
+          pip install --upgrade pep8-naming
+    - name: Setup Homebrew packages
+      run: |
+        brew update
+        brew upgrade
+        brew install gcc gnupg2 dash kcov
```
Comment (ax3l) [inline .github/workflows/macos_unit_tests.yaml:35]: Maybe try: https://github.com/kelaberetiv/TagUI/issues/86
```suggestion
        brew install gcc gnupg2 dash kcov openssl
```

Comment (alalazo) [reply .github/workflows/macos_unit_tests.yaml:35]: It didn't work here 6e22abe65c64dc4ca42e1161c38e0b345b2a91c4 but there are also weird interactions with the order in which `pip` and `brew` are called so I'll give it a try again.

**Comment by alalazo** [comment]
@tgamblin @ax3l Ready for a review!

**Pull request merged**